src/backend.rs

```diff
@@ -75,31 +77,146 @@ impl Registers {
     // do on-the-fly register allocation here.
+    ctx.sp_depth += 1;
     dynasm!(ops
         ; push Rq(gpr)
     );
-    regs.release_scratch_gpr(gpr);
+    ctx.regs.release_scratch_gpr(gpr);
 }
 
-fn pop_i32(ops: &mut Assembler, regs: &mut Registers) -> GPR {
-    let gpr = regs.take_scratch_gpr();
+fn pop_i32(ctx: &mut Context, ops: &mut Assembler) -> GPR {
+    ctx.sp_depth -= 1;
+    let gpr = ctx.regs.take_scratch_gpr();
     dynasm!(ops
         ; pop Rq(gpr)
     );
     gpr
 }
 
-pub fn add_i32(ops: &mut Assembler, regs: &mut Registers) {
-    let op0 = pop_i32(ops, regs);
-    let op1 = pop_i32(ops, regs);
+pub fn add_i32(ctx: &mut Context, ops: &mut Assembler) {
+    let op0 = pop_i32(ctx, ops);
+    let op1 = pop_i32(ctx, ops);
     dynasm!(ops
         ; add Rq(op0), Rq(op1)
     );
-    push_i32(ops, regs, op0);
-    regs.release_scratch_gpr(op1);
+    push_i32(ctx, ops, op0);
+    ctx.regs.release_scratch_gpr(op1);
+}
+
+fn sp_relative_offset(ctx: &mut Context, slot_idx: u32) -> i32 {
+    ((ctx.sp_depth as i32) + slot_idx as i32) * WORD_SIZE as i32
+}
+
+pub fn load_i32(ctx: &mut Context, ops: &mut Assembler, local_idx: u32) {
+    let gpr = ctx.regs.take_scratch_gpr();
+    let offset = sp_relative_offset(ctx, local_idx);
+    dynasm!(ops
+        ; mov Rq(gpr), [rsp + offset]
+    );
+    push_i32(ctx, ops, gpr);
+}
+
+pub fn store_i32(ctx: &mut Context, ops: &mut Assembler, local_idx: u32) {
+    let gpr = pop_i32(ctx, ops);
+    let offset = sp_relative_offset(ctx, local_idx);
+    dynasm!(ops
+        ; mov [rsp + offset], Rq(gpr)
+    );
+    ctx.regs.release_scratch_gpr(gpr);
+}
+
+pub fn prepare_return_value(ctx: &mut Context, ops: &mut Assembler) {
+    let ret_gpr = pop_i32(ctx, ops);
+    if ret_gpr != RAX {
+        dynasm!(ops
+            ; mov Rq(RAX), Rq(ret_gpr)
+        );
+        ctx.regs.release_scratch_gpr(ret_gpr);
+    }
+}
+
+pub fn copy_incoming_arg(ctx: &mut Context, ops: &mut Assembler, arg_pos: u32) {
+    let loc = abi_loc_for_arg(arg_pos);
+
+    // First, ensure the argument is in a register.
+    let reg = match loc {
+        ArgLocation::Reg(reg) => reg,
+        ArgLocation::Stack(offset) => {
+            // RAX is always scratch?
+            dynasm!(ops
+                ; mov Rq(RAX), [rsp + offset]
+            );
+            RAX
+        }
+    };
+
+    // And then move a value from a register into local variable area on the stack.
+    let offset = sp_relative_offset(ctx, arg_pos);
+    dynasm!(ops
+        ; mov [rsp + offset], Rq(reg) 
+    );
+}
+
+pub fn prologue(_ctx: &mut Context, ops: &mut Assembler, stack_slots: u32) {
+    let framesize: i32 = stack_slots as i32 * WORD_SIZE as i32;
+    dynasm!(ops
+        ; push rbp
+        ; mov rbp, rsp
+        ; sub rsp, framesize
+    );
+}
+
+pub fn epilogue(_ctx: &mut Context, ops: &mut Assembler) {
+    dynasm!(ops
+        ; mov rsp, rbp
+        ; pop rbp
+        ; ret
+    );
```

Comment:
Another benefit of tracking `sp_depth` is that we should be able to assert that it gets back down to 0 at the end, after subtracting off the frame size. This can often catch imbalanced pushes and pops.